scalafix-rules/src/main/scala/scalafix/internal/pc/Embedded.scala

```diff
@@ -0,0 +1,106 @@
+package scalafix.internal.pc
+
+import java.net.URLClassLoader
+import java.nio.file.Path
+import java.util.ServiceLoader
+
+import scala.jdk.CollectionConverters._
+
+import scala.meta.pc.PresentationCompiler
+
+import coursierapi.Dependency
+import coursierapi.Fetch
+import coursierapi.MavenRepository
+
+object Embedded {
+
+  def presentationCompiler(
+      scalaVersion: String
+  ): PresentationCompiler = {
+    val deps =
+      scala3PresentationCompilerDependencies(scalaVersion)
+    val jars = Fetch
+      .create()
+      .addDependencies(deps: _*)
+      .addRepositories(
+        MavenRepository.of(
+          "https://oss.sonatype.org/content/repositories/snapshots"
+        )
+      )
+      .fetch()
+      .asScala
+      .map(_.toPath())
+      .toSeq
+    val classloader = newPresentationCompilerClassLoader(jars)
+
+    val presentationCompilerClassname =
+      if (supportPresentationCompilerInDotty(scalaVersion)) {
+        "dotty.tools.pc.ScalaPresentationCompiler"
+      } else {
+        "scala.meta.pc.ScalaPresentationCompiler"
+      }
+
+    serviceLoader(
+      classOf[PresentationCompiler],
+      presentationCompilerClassname,
+      classloader
+    )
+  }
+
+  private def supportPresentationCompilerInDotty(scalaVersion: String) = {
+    scalaVersion
+      .replaceAll(raw"-RC\d+", "")
+      .split("\\.")
+      .take(3)
+      .map(_.toInt) match {
+      case Array(3, minor, patch) => minor > 3 || minor == 3 && patch >= 4
+      case _ => false
+    }
+  }
+
+  private def scala3PresentationCompilerDependencies(version: String) =
+    if (supportPresentationCompilerInDotty(version)) {
+      val dep = Dependency
+        .of("org.scala-lang", "scala3-presentation-compiler_3", version)
```

Comment:
As a follow up to my comments about repositories, I am trying to understand what forces us to have that dynamic coursier resolution which brings a lot of complexity. I guess we need a PC matching as much as possible the compiler used by the client? Would it be an option to cross-build scalafix_rules to each minor version of Scala 3 to have build-time dependencies instead?

Reply:
> I guess we need a PC matching as much as possible the compiler used by the client?

Ideally yes, the compiler API itself is not binary compatible, so we need to have the same version of PC and the compiler and if possible also the same Scala version used by the user. For Scala 2 we are kind of cheating, since we should be doing the same thing. For example some bug could be fixed in a later version and if someone has an earlier one the rule might break things for them (granted I haven't seen too many issues like this)

> Would it be an option to cross-build scalafix_rules to each minor version of Scala 3 to have build-time dependencies instead

It would be possible for sure, but might be a lot of effort and we would need to release every time a new release is out, which this would avoid (we can still keep dynamic resolution as fallback in that case). We could also release rules for each minor, which would be a bit less work for sure.
Possibly we could

Reply:
> We could also release rules for each minor

My assumption here is that the 3.x.0 PC is able to handle what the 3.x.y compiler can. Is that the case?

Reply:
I think you're right. Maybe I am being too cautious? We can start with releasing for LTS and Next.

Reply:
Using LTS for 3.4 can be problematic since new source-level features won't be handled. But maybe we can build scalafix-rules_3 with latest (and scalafix-core_3 with LTS for maximum compatibility for external rule authors), and always use that? That would be very similar to what we do for 2.12 and 2.13 actually.

Reply:
ok, let's do that. So scalafix-rules_3 to be build with 3.3.x, 3.4.x and 3.5.x and core just for 3.3.x, right?

Reply:
I would say: build scalafix-rules_3 with NEXT (to statically bring the most powerful/up-to-date PC, that should be able to handle bytecode/TASTY compiled by a prior version), and scalafix-core with LTS (for rule authors to be flexible).

Reply:
We can think of a more dynamic solution later (and how to honour custom repositories), which would allow scalafix to handle future/upcoming Scala 3 minor versions, but it has been this way for Scala 2.x so users won't find it weird to update as Scala versions come out.

Reply:
The only thing I am wondering if the type could change in between major version (I think it can) and in the case using latest PC under Next could break your code which might be under LTS

Reply:
Oh I see, the latest Next PC would infer a type that LTS would not support 😞  I am not sure how common that use case is, and whether it justifies the complexity at build time and also at resolution time, since we'd have to suffix artefacts ourselves as I don't think coursier (the one already existing in scalafix-interfaces) supports "3.x" lookups?

Reply:
At a minimum we need to publish for 3.3, 3.4, 3.5 etc. since we can get changes that even remove certain syntax, so PC would no longer be able to work with that. Sorry for the back and forth but I am doing a lot of pondering here.

My opinion about final implementation:
- core published with LTS
- rules published for each minor to be on the safe side -> not sure how to solve this build wise, maybe full compat needed
- leave fallback for anything that we don't know about -> that should be safe enough as the alternative for people is that it will not work.

Reply:
>  since we'd have to suffix artefacts ourselves as I don't think coursier (the one already existing in scalafix-interfaces) supports "3.x" lookups?

We would need to do a full binary suffix and release one for each minor. So now we would have `rules_3.3.4` for 3.3 line, `rules_3.4.2` for 3.4 line and `rules_3.5.0` for 3.5 line.

It might be easier at this point to deal with coursier resolution :sweat_smile:

Reply:
scalafix-cli (whic brings in scalafix-rules) is fully cross-built already, so that wouldn't be a problem at least.

What I am concerned about is that [ScalafixInterface  only asks for a binary version](https://github.com/scalacenter/scalafix/blob/5a02480f5ee126a3804ac3bbd6e3d13cb864b02a/scalafix-interfaces/src/main/java/scalafix/interfaces/Scalafix.java#L105) to know which scalafix-cli to load, so it looks like we would have to update all clients (which rely on scalafix-interfaces only, example [sbt-scalafix](https://github.com/scalacenter/sbt-scalafix/blob/9bf8b8b1b21ad098ef75995622b20f263dd050fb/src/main/scala/scalafix/sbt/ScalafixPlugin.scala#L372-L374)) to pass more information than just "3".

I will try to reflect on the tradeoffs...

Reply:
Couldn't we basically require 3.3 etc ? That wouldn't be a super huge change for the clients (just throw an exception let them know) and we can translate it to the known artifacts at the release of the interface. If the interface doesn't know that version, it just loads LTS, which resolves things dynamically as a last resort?

Reply:
Right, it wouldn't be a breaking change anyway since the full version is not needed for 2.x, and that's the only thing supported by existing clients. Adding this to https://github.com/scalacenter/scalafix/issues/1680.